January 2019 DAC codelist updates

andylolz opened the pull request:
These updates are all from the DAC XML source, available here:

https://webfs.oecd.org/crs-iati-xml/Lookup/

This replaces #249.

---

To summarise the changes here:

# Aid Type

### Codes added:
 * `H03` - Asylum-seekers ultimately accepted
 * `H04` - Asylum-seekers ultimately rejected
 * `H05` - Recognised refugees

# Sector Category

### Codes added:
 * `123` - Non-communicable diseases (NCDs)

# Sector

### Codes added:
 * `11250` - School feeding
 * `12310` - NCDs control, general
 * `12320` - Tobacco use control
 * `12330` - Control of harmful use of alcohol and drugs
 * `12340` - Promotion of mental health and well-being
 * `12350` - Other prevention and treatment of NCDs
 * `12382` - Research for prevention and control of NCDs
 * `15190` - Facilitation of orderly, safe, regular and responsible migration and mobility
 * `16070` - Labour Rights
 * `16080` - Social Dialogue
 * `24050` - Remittance facilitation, promotion and optimisation
 * `25030` - Business development services
 * `25040` - Responsible Business Conduct
 * `43060` - Disaster Risk Reduction
 * `43071` - Food security policy and administrative management
 * `43072` - Household food security programmes
 * `43073` - Food safety and quality
 * `74020` - Multi-hazard response preparedness
 * `93011` - Refugees/asylum seekers in donor countries - food and shelter
 * `93012` - Refugees/asylum seekers in donor countries - training
 * `93013` - Refugees/asylum seekers in donor countries - health
 * `93014` - Refugees/asylum seekers in donor countries - other temporary sustenance
 * `93015` - Refugees/asylum seekers in donor countries - voluntary repatriation
 * `93016` - Refugees/asylum seekers in donor countries - transport
 * `93017` - Refugees/asylum seekers in donor countries - rescue at sea
 * `93018` - Refugees/asylum seekers in donor countries - administrative costs

### Codes withdrawn:
 * `41050` - Flood prevention/control
 * `74010` - Disaster prevention and preparedness

# Finance Type Category

### Codes added:
 * `0` - NON FLOW ITEMS

# Finance Type

### Codes added:
 * `1` - GNI: Gross National Income
 * `2` - ODA % GNI
 * `3` - Total Flows % GNI
 * `4` - Population

andylolz commented:
Related IATI discuss post:
https://discuss.iatistandard.org/t/january-2019-dac-codelist-updates/1622

samuele-mattiuzzo commented:
@andylolz for the BAs benefit, would you mind providing a diff between your import and the original XML file (if you have something like that available that is)? Do you use the script you suggested we'd use in another PR?

Thank you!

andylolz commented:
> Do you use the script you suggested we'd use in another PR?

Nope – I did this manually :) The script in #172 processes the DAC Excel file (well, it processes [some CSV on datahub.io](https://datahub.io/core/dac-and-crs-code-lists), but that comes from the DAC Excel file). @bill-anderson has said he doesn’t want to use the Excel file (more sustainable solution etc) so this PR uses the XML instead.

> would you mind providing a diff between your import and the original XML file (if you have something like that available that is)?

A diff is maybe tricky because the DAC XML file ([available here](https://webfs.oecd.org/crs-iati-xml/Lookup/)) is just one big file. But I can explain the steps I went through. I did the following bits of cleanup:

 * Pretty print (with 4 space indentation)
 * Remove DAC namespace elements, because these aren’t valid
 * Remove trailing whitespace in text nodes
 * Replace bad `@status`es with either "active" or "withdrawn" (The `@status` attribute in the DAC XML sometimes contains "active-MCD", "active- Pilot" or "Vonlontary basis" [sic], which are not valid statuses. I’ve flagged this issue with your colleagues and with Valerie by email, but for the purposes of this PR I’ve fixed these manually.)
 * Split into constituent files
 * Add the IATI metadata back (e.g. DAC calls the channel code codelist "Channelcode", whereas IATI calls it "CRSChannelCode")

I think that’s everything. Here’s what I haven’t done:

 * Reordered codes to match the previous order (which would probably make diffs a bit easier to read)
 * Gone through and checked for removed codes (anything removed is a problem, since it should instead be marked as `status="withdrawn"`)

The diff in this PR shows that quite a lot of stuff has changed. I guess that’s mostly because the source has changed from Excel to XML. I think it will be difficult to verify and merge this PR for that reason. If the goal is to eventually use XML from the DAC as the source for these replicated codelists, then I’d be tempted to go back to the DAC technical team with a list of stuff to fix at their end, and use the Excel file as the source in the interim.

---

I’m very pleased you’re looking at this, because it’s really important that these replicated codelists are kept in sync with source. For instance, a validator might say that a dataset is invalid because a bad sector code is used, when in fact the problem might be that the IATI replicated Sector codelist is out of sync, and doesn’t include a complete list of sector codes. A publisher could also be scored down on the Aid Transparency Index for the same reason. Or an aid management system might rely on these codelists for interpreting published IATI data.

Anyway – I’d be happy to discuss next steps.

samuele-mattiuzzo commented:
@andylolz fab, thanks! Petya and the BAs have this to check on their todo list, it'll be checked during this week!

PetyaKangalova commented:
We have now received a response from the OECD that the XML files has been updated and both Excel and XML files have been pulled from the same source.

> Both xml and xl file have been regenerated (from SQL as unique source, except for Channel codes) and are available on our website http://www.oecd.org/dac/financing-sustainable-development/development-finance-standards/dacandcrscodelists.htm.

From a quick look of the difference I have identified before the codelists are now identical in the Excel and XML files so I think we can use the updated XML to update the codelists on the  IATI website.

@andylolz  Is there a way of easily re-doing what you have done so far with the updated XML file? Then I can review the pull request.  If it requires a lot of manual work for you, then I can look into making the comparison and adding the pull requests.

andylolz commented:
@PetyaKangalova no problem – I’ll try and get this sorted today.

PetyaKangalova commented:
Thanks @andylolz ! I am off on Monday and in meetings all of Tuesday but should be able to review mid-next week! Thanks again!

andylolz commented:
Okay – PR updated using the latest (updated) version of DAC XML. I followed the same steps described above.

PetyaKangalova commented:
@andylolz thanks again for redoing the commit. Really appreciate it!  It took me a while to review all the changes as there are quite a lot of them! See summary below:

- **Aid Type Category**- ready to approve changes
- **Aid Type**- ready to approve changes
- **CRS Channel Code**-  Valerie confirmed that CRS Channel code is the only one not created from source. In the XML you have used some existing codes are missing and don’t think that is on purpose
- **Collaboration Type**- ready to approve changes
- **Finance Type Category**-  ready to approve changes
- **Finance Type**- ready to approve changes
- **Flow Type-** ready to approve changes
- **Sector**- ready to approve changes except for code 74010 and  74020
- **Sector Category**- need to understand why descriptions have been removed before approving


**Next steps:**

1. @andylolz , would you be able to remove the CRS Channel Code section from the pull request for now? I will contact Valerie to get confirmation but feel like this one will take some time as the XML has not been created from their source database and don’t want to hold the other changes.
2. I will contact Valerie to get confirmation on whether sector code 74010 has been withdrawn and why description for sector categories have been removed.
3. Once I get confirmation on 2 , I will make the necessary changes and approve the pull request. As I was reviewing the changes I kept track of all of them (whether it was code addition or change of name or description). I will then work on adding all changes to the [non-embedded codelist changelog ](http://reference.iatistandard.org/203/upgrades/nonembedded-codelist-changelog/)(just for the sector codelist there are more than 40 changes so might take us some time)
4. Once codelist changes and changelog have been approved and deployed, we will add a post on IATI Discuss.
5. We will also contact publishing tool providers to make them aware of the changes.

andylolz commented:
> 1. @andylolz , would you be able to remove the CRS Channel Code section from the pull request for now?

Kk, done.

> * **Sector**- ready to approve changes except for code 74010 and  74020

Oh, good spot! The same applies to 41050 (Flood prevention/control), which has also disappeared.

Also, the following withdrawn sector codes have disappeared:

 * 15120: Public sector financial management
 * 15140: Government administration
 * 15161: Elections
 * 15162: Human rights
 * 15163: Free flow of information
 * 15164: Women's equality organisations and institutions
 * 23010: Energy policy and administrative management
 * 23020: Power generation/non-renewable sources
 * 23030: Power generation/renewable sources
 * 23040: Electrical transmission/ distribution
 * 23050: Gas distribution
 * 23061: Oil-fired power plants
 * 23062: Gas-fired power plants
 * 23063: Coal-fired power plants
 * 23064: Nuclear power plants
 * 23065: Hydro-electric power plants
 * 23066: Geothermal energy
 * 23067: Solar energy
 * 23068: Wind power
 * 23069: Ocean power
 * 23070: Biomass
 * 23081: Energy education/training
 * 23082: Energy research
 * 92010: Support to national NGOs
 * 92020: Support to international NGOs
 * 92030: Support to local and regional NGOs

They may have been replaced by other codes, but the idea is they’re supposed to remain in perpetuity as `status="withdrawn"`.

andylolz commented:
> 3\. just for the sector codelist there are more than 40 changes so might take us some time

I’ve mentioned elsewhere that I’m in favour of scrapping this changelog. I’m unconvinced it’s worth your time. It wasn’t updated for the last DAC codelist update (see: IATI/IATI-Guidance#312) so it’s only a partial list of changes anyway.

> 5\. We will also contact publishing tool providers to make them aware of the changes.

Okay – this is very generous of you, but again I don’t think this should be standard practice. Tool providers should be keeping an eye on discuss, or routinely pulling from source. That’s the system as documented. If they start relying on updates from you then that just becomes an extra overhead for you.

PetyaKangalova commented:
@andylolz 

> Kk, done.
 
Thank you!

>Oh, good spot! The same applies to 41050 (Flood prevention/control), which has also disappeared.

Thank you for flagging. I missed this one!

>Also, the following withdrawn sector codes have disappeared:

* 15120: Public sector financial management
* 15140: Government administration
* 15161: Elections
* 15162: Human rights
* 15163: Free flow of information
* 15164: Women's equality organisations and institutions
* 23010: Energy policy and administrative management
* 23020: Power generation/non-renewable sources
* 23030: Power generation/renewable sources
* 23040: Electrical transmission/ distribution
* 23050: Gas distribution
* 23061: Oil-fired power plants
* 23062: Gas-fired power plants
* 23063: Coal-fired power plants
* 23064: Nuclear power plants
* 23065: Hydro-electric power plants
* 23066: Geothermal energy
* 23067: Solar energy
* 23068: Wind power
* 23069: Ocean power
* 23070: Biomass
* 23081: Energy education/training
* 23082: Energy research
* 92010: Support to national NGOs
* 92020: Support to international NGOs
* 92030: Support to local and regional NGOs

Yes, I agree! I also noticed that there were a few new  'withdrawn' as of 2015 that were not on the IATI list. As they are already withdrawn I was not so concerned but it means the XML is not consistent.

On your point for the changelog I agree that it is a lot of effort. However, this time round there are quite a lot of new codes and it will be important to alert people which ones those are and also make sure organisations can start using them via the various publishing tools. Hence, dropping them a quick email to speed up the process, but it is indeed their responsibility of the tool providers to keep them up-to-date.

Waiting to hear from Valerie and will then action the changes!

andylolz commented:
> Yes, I agree! I also noticed that there were a few new 'withdrawn' as of 2015 that were not on the IATI list. As they are already withdrawn I was not so concerned but it means the XML is not consistent.

Yes that’s true, but I’d expect DAC to have a better record of withdrawn codes than IATI (since IATI only started recording these relatively recently). So withdrawn codes in the XML that were not previously known to IATI are probably a good thing :)

PetyaKangalova commented on xml/Sector.xml:
```diff
@@ -3090,6 +3102,18 @@
             <category>520</category>
         </codelist-item>
+        <codelist-item status="withdrawn" activation-date="2018-01-01">
+            <code>52020</code>
```
I am not sure why this was added as withdrawn- I cannot find in our existing codelist and seems to be withdrawn relatively recently. I will check with Valerie whether it was done in error. It will not have implications for user as it is withdrawn anyway, but it is strange.

andylolz replied:
[This document](https://one.oecd.org/document/DCD/DAC/STAT(2018)40/REV2/en/pdf) (from November 2018) says: 

> The code `52020` “Household food security programmes” previously approved at the July 2017 WP-STAT meeting but not yet in effect, is moved under the “other multisector” category (`430`) and assigned a new code `43072`.

You can download a copy of the XLS as it stood in August 2017 from here:
https://github.com/datasets/dac-and-crs-code-lists/blob/0a354cefc5123cafd2b79aecb2cc31bb9753f9a9/source/codelists.xls

It doesn’t include the code, but perhaps “not yet in effect” means it didn’t actually ever make it onto the codelist. Not sure.

andylolz replied:
BTW I tried to use the dashboard to check whether `52020` is used in IATI data, but found a bug 😞

PetyaKangalova replied:
Thanks @andylolz  I just got the response a response from Valerie which is in line with the comment you have provided. I think this code was never added to the DAC Excel spreadsheet as there is normally a timelag from approval until addition to the list, hence it was never replicated.

>The code 52020 is a special case. Code 52020 has been approved in the context of aligning purpose codes to SDGs  in the nutrition area in 2017 (DCD/DAC/STAT/RD(2017)11), to be implemented in 2019 on 2018 flows. But in the context of revision of purpose codes in light of SDGs in food safety and food security area, this purpose code has been recoded in 43072 (DCD/DAC/STAT(2018)40/REV2) also to be implemented in 2019 on 2018 flows. 
We have decided to stop to disseminate it as it is confusing.

As such, I don't think we need to be adding it. 

About the Dashboard, I was also unable to identify the publishers using this code.

andylolz commented on xml/FinanceType-category.xml:
```diff
@@ -9,127 +9,147 @@
+        <codelist-item status="active">
+            <code>0</code>
+            <name>
+                <narrative>NON FLOW ITEMS</narrative>
```
I have a feeling this code is deliberately excluded from the IATI replicated codelist. @bill-anderson can confirm or deny.

samuele-mattiuzzo replied:
@andylolz @PetyaKangalova is this above comment here holding up the merge or can it still go?

andylolz replied:
Aha – here’s why I think this is deliberately excluded: https://github.com/IATI/IATI-Codelists-NonEmbedded/issues/16#issuecomment-51773027

andylolz replied:
I have no idea whether that means they should or shouldn’t be included. I shall leave it with you to decide!

PetyaKangalova replied:
I would agree that finance type '0- non flow items' does not fit within the activity standard as it does not describe type of finance at activity level.  It is much wider than that. I would propose that we do **not** replicate- waiting for confirmation from @bill-anderson so  @samuele-mattiuzzo please hold off merging until then.

PetyaKangalova replied:
I just talked to @bill-anderson  and the view is that we keep this code in, the reason being that we replicate all 'OECD DAC' codelists exactly as they are in the table provided by the DAC.  @samuele-mattiuzzo  this means we can continue merging this request.

However, what will be useful is to add a note in the Finance Type and Finance Type (Category) to note that non-flow items are not activity specific and we do not expect the codes to be used when reporting finance types in the IATI activity standard.

andylolz replied:
Not really sure about the note idea.

If it’s important enough that it’s a problem, then you can either add a rule to the ruleset, or leave the codes out.

I guess in general, I think it’s best to avoid situations where the standard strongly advises something, but doesn’t enforce in any way.

bill-anderson replied:
re: "or leave the codes out."
As already mentioned the standard replicates third-party code lists. This is a well-established principle.

andylolz replied:
Hi @bill-anderson – yes, but cf. this comment from 2014 @bill-anderson:
https://github.com/IATI/IATI-Codelists-NonEmbedded/issues/16#issuecomment-51773027

andylolz commented on xml/FinanceType.xml:
```diff
@@ -9,7 +9,19 @@
+                <narrative/>
+                <narrative xml:lang="fr"/>
+            </description>
+            <category>0</category>
```
I guess any FinanceType codes with category 0 probably shouldn’t be added, for the same reason as https://github.com/IATI/IATI-Codelists-NonEmbedded/pull/283/files#r260298899

andylolz replied:
Marking as resolved, as per Petya’s comment here: https://github.com/IATI/IATI-Codelists-NonEmbedded/pull/283#discussion_r

andylolz commented:
I’ve added a summary of changes in the PR description.

andylolz commented:
Seems like the conclusion is: Merge this, and then add a note to the FinanceType codelist in a new pull request.

@PetyaKangalova is that right?

andylolz commented:
:star: :star2: :star:
